tensorrt_llm/_torch/modules/attention.py

```diff
@@ -191,28 +186,8 @@ def forward(
         if self.apply_rotary_emb and position_ids is not None:
             q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size],
                                 dim=-1)
-            if hasattr(self, 'q_norm') and hasattr(self, 'k_norm'):
-                # Add qk-norm
-                if hasattr(self, 'ln_events'):
-                    q_l2norm = lambda: self.q_norm(q.reshape(-1, self.head_dim)
-                                                   ).reshape(-1, self.q_size)
-                    k_l2norm = lambda: self.k_norm(k.reshape(-1, self.head_dim)
-                                                   ).reshape(-1, self.kv_size)
-                    q, k = maybe_execute_in_parallel(
-                        q_l2norm,
-                        k_l2norm,
-                        self.ln_events[0],
-                        self.ln_events[1],
-                        self.aux_stream,
-                    )
-                else:
-                    q_by_head = q.reshape(-1, self.head_dim)
-                    q_by_head = self.q_norm(q_by_head)
-                    q = q_by_head.view(q.shape)
-                    k_by_head = k.reshape(-1, self.head_dim)
-                    k_by_head = self.k_norm(k_by_head)
-                    k = k_by_head.view(k.shape)
-
+            if self.use_qk_norm:
+                q, k = self.apply_qk_norm(q, k)
```

Comment:
Here you apply `qk_norm` first and then `apply_rotary`, which is for QWen3, I suppose. But if you look at the llama4 code, the order is the opposite. You do `apply_rotary` first, and then `qk_norm`.

I only noticed this while reviewing the PR today. Maybe it's not the right design to try to implement `qk_norm` generically.